src/Microsoft.Data.Analysis/Computations/Arithmetic.cs

```diff
@@ -0,0 +1,328 @@
+﻿
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// Generated from Arithmetic.tt. Do not modify directly
+
+using System;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Data.Analysis
+{
+    ////////////////////////////////////////
+    //Factory Class                       //
+    ////////////////////////////////////////
+
+    internal static class Arithmetic
+    {
+        public static IArithmetic<T> GetArithmetic<T>()
+            where T : unmanaged
+        {
+            if (typeof(T) == typeof(bool))
+                return (IArithmetic<T>)new BoolArithmetic();
+            else if (typeof(T) == typeof(byte))
+                return (IArithmetic<T>)new ByteArithmetic();
+            else if (typeof(T) == typeof(char))
+                return (IArithmetic<T>)new CharArithmetic();
+            else if (typeof(T) == typeof(decimal))
+                return (IArithmetic<T>)new DecimalArithmetic();
+            else if (typeof(T) == typeof(double))
+                return (IArithmetic<T>)new DoubleArithmetic();
+            else if (typeof(T) == typeof(float))
+                return (IArithmetic<T>)new FloatArithmetic();
+            else if (typeof(T) == typeof(int))
+                return (IArithmetic<T>)new IntArithmetic();
+            else if (typeof(T) == typeof(long))
+                return (IArithmetic<T>)new LongArithmetic();
+            else if (typeof(T) == typeof(sbyte))
+                return (IArithmetic<T>)new SByteArithmetic();
+            else if (typeof(T) == typeof(short))
+                return (IArithmetic<T>)new ShortArithmetic();
+            else if (typeof(T) == typeof(uint))
+                return (IArithmetic<T>)new UIntArithmetic();
+            else if (typeof(T) == typeof(ulong))
+                return (IArithmetic<T>)new ULongArithmetic();
+            else if (typeof(T) == typeof(ushort))
+                return (IArithmetic<T>)new UShortArithmetic();
+            else if (typeof(T) == typeof(DateTime))
+                return (IArithmetic<T>)new DateTimeArithmetic();
+            throw new NotSupportedException();
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref Vector<T> AsVector<T>(ref T start, int offset)
+            where T : struct => ref Unsafe.As<T, Vector<T>>(ref Unsafe.Add(ref start, offset));
```

Comment:
To avoid any potential alignment failure and increase portability this shouldn't use `Unsafe.As` and use `Unsafe.ReadUnaligned` instead.

PS: I see this is the genarted file, but easier to comment on here

Reply:
Do you suggest to use `Unsafe.ReadUnaligned<Vector<T>>(ref Unsafe.As<T, byte>(ref Unsafe.Add(ref start, offset)));` instead? The problem is that it returns a value, not a reference and in this code reference is used to write a result of the calculation to the destination span.

I looked inside ReadUnaligned implementation, it's look quite similar: `return As<byte, T>(ref Unsafe.AsRef(in source));`, could you please explain why it's better?

Reply:
Actually I took this approach from .Net System.Numerics.Tensors library: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netstandard.cs#L745

Reply:
@stephentoub could you please advise on this matter?

Reply:
> in this code reference is used

Ah, I didn't spot the `ref` here. Sorry.
W/o the `ref` then Unsafe.ReadUnaligned is preferable, because of the explicit "unaligned" hint by which the JIT will emit the correct loads to prevent any alignment failures (to which some types of hardware are sensible and my fault).